File: src/aws/lambdas/incremental_distribution/src/test/java/uk/nhs/nhsx/diagnosiskeydist/ZipFileUtility.java
package uk.nhs.nhsx.diagnosiskeydist;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

public class ZipFileUtility {

    private static final String TEMP_DIR = "DistributionServiceTest";

    public static File extractZipFileToTempLocation(File file) throws IOException {
        File destDir = Files.createTempDirectory(TEMP_DIR).toFile();

        byte[] buffer = new byte[1024];
        ZipInputStream zipStream = new ZipInputStream(new FileInputStream(file));
        ZipEntry zipEntry = zipStream.getNextEntry();
        while (zipEntry != null) {
            File newFile = new File(destDir, zipEntry.getName());
            FileOutputStream outputStream = new FileOutputStream(newFile);
            int length;
            while ((length = zipStream.read(buffer)) > 0) {
                outputStream.write(buffer, 0, length);
            }
            outputStream.close();
            zipEntry = zipStream.getNextEntry();
        }
        zipStream.close();

        return destDir;
    }
}

File: src/aws/lambdas/incremental_distribution/src/test/java/uk/nhs/nhsx/diagnosiskeydist/ZipFileUtility.kt
package uk.nhs.nhsx.diagnosiskeydist

import net.lingala.zip4j.ZipFile
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.nio.file.Files
import java.util.zip.ZipInputStream

object ZipFileUtility {
    private const val TEMP_DIR = "DistributionServiceTest"

    fun extractZipFileToTempLocation(file: File): File {
        val destDir = Files.createTempDirectory(TEMP_DIR).toFile()
        val buffer = ByteArray(1024)
        val zipStream = ZipInputStream(FileInputStream(file))
        var zipEntry = zipStream.nextEntry
        while (zipEntry != null) {
            val newFile = File(destDir, zipEntry.name)
            val outputStream = FileOutputStream(newFile)
            var length: Int
            while (zipStream.read(buffer).also { length = it } > 0) {
                outputStream.write(buffer, 0, length)
            }
            outputStream.close()
            zipEntry = zipStream.nextEntry
        }
        zipStream.close()
        return destDir
    }

    fun extractZipFileToTempLocation(file: File, password:String): File {
        val destDir = Files.createTempDirectory(TEMP_DIR).toFile()

        val zipFile = ZipFile (file)
        if (zipFile.isEncrypted) {
            zipFile.setPassword(password.toCharArray())
        }
        zipFile.extractAll(destDir.absolutePath)
        return destDir
    }
}
